physical/mysql.go

```diff
@@ -0,0 +1,204 @@
+	}
+
+	// Create the required table if it doesn't exists.
+	create_query := "CREATE TABLE IF NOT EXISTS " + database + "." + table + " (vault_key varchar(512), vault_value mediumblob, PRIMARY KEY (vault_key))"
+	create_stmt, err := db.Prepare(create_query)
+	if err != nil {
+		return nil, MySQLPrepareStmtFailure
+	}
+	defer create_stmt.Close()
+
+	_, err = create_stmt.Exec()
+	if err != nil {
+		return nil, MySQLExecuteStmtFailure
+	}
+
+	// Map of query type as key to prepared statement.
+	statements := make(map[string]*sql.Stmt)
+
+	// Prepare statement for put query.
+	insert_query := "INSERT INTO " + database + "." + table + " VALUES( ?, ? ) ON DUPLICATE KEY UPDATE vault_value=VALUES(vault_value)"
+	insert_stmt, err := db.Prepare(insert_query)
+	if err != nil {
+		return nil, MySQLPrepareStmtFailure
+	}
+	statements["put"] = insert_stmt
+
+	// Prepare statement for select query.
+	select_query := "SELECT vault_value FROM " + database + "." + table + " WHERE vault_key = ?"
+	select_stmt, err := db.Prepare(select_query)
+	if err != nil {
+		return nil, MySQLPrepareStmtFailure
+	}
+	statements["get"] = select_stmt
+
+	// Prepare statement for delete query.
+	delete_query := "DELETE FROM " + database + "." + table + " WHERE vault_key = ?"
+	delete_stmt, err := db.Prepare(delete_query)
+	if err != nil {
+		return nil, MySQLPrepareStmtFailure
+	}
+	statements["delete"] = delete_stmt
+
+	// Setup the backend.
+	m := &MySQLBackend{
+		client:     db,
+		table:      table,
+		database:   database,
+		statements: statements,
+	}
+
+	return m, nil
+}
+
+// Put is used to insert or update an entry.
+func (m *MySQLBackend) Put(entry *Entry) error {
+	defer metrics.MeasureSince([]string{"mysql", "put"}, time.Now())
+
+	_, err := m.statements["put"].Exec(entry.Key, entry.Value)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Get is used to fetch and entry.
+func (m *MySQLBackend) Get(key string) (*Entry, error) {
+	defer metrics.MeasureSince([]string{"mysql", "get"}, time.Now())
+
+	var result []byte
+
+	err := m.statements["get"].QueryRow(key).Scan(&result)
+	if err != nil {
+		return nil, nil
+	}
+
+	// Handle a non-existing value
+	if result == nil {
+		return nil, nil
+	}
+
+	ent := &Entry{
+		Key:   key,
+		Value: result,
+	}
+
+	return ent, nil
+}
+
+// Delete is used to permanently delete an entry
+func (m *MySQLBackend) Delete(key string) error {
+	defer metrics.MeasureSince([]string{"mysql", "delete"}, time.Now())
+
+	_, err := m.statements["delete"].Exec(key)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// List is used to list all the keys under a given
+// prefix, up to the next prefix.
+func (m *MySQLBackend) List(prefix string) ([]string, error) {
+	defer metrics.MeasureSince([]string{"mysql", "list"}, time.Now())
+
+	// Query to get all keys matching a prefix.
+	list_query := "SELECT vault_key FROM " + m.database + "." + m.table + " WHERE vault_key LIKE '" + prefix + "%'"
```

Comment:
@armon I am unable to figure out how to do mysql escaping. Can you provide me link of some example. Sorry to bother you.